### ISSUE TITLE
Configurable Properties

### ISSUE THREAD

**Issue by lyes-aknouche** [ot4i/node-for-log4j#7]
Hello, first of all, thank you for this great plugin!

I'm currently working on migrating ACE applications using this plugin version 1.2 to version 2.0.1.
It seems that configurable properties have been removed in version 2.0.1 and 2.0.0 (configurable properties were available in version 1.1 / 1.2).

Would it be possible to re-introduce configurable properties for the log4j ACE node?

After digging the files in Log4jLoggingPluginFeature_v2.0.1.zip, it seems that the below files may need to be updated:
- **plugins\Log4jLoggingPlugin_2.0.0\Log4jLoggingPlugin.msgnode**
This file used to have (in versions 1.1 / 1.2) the attributes **configurable="true"** for each XML element **propertyDescriptor**
- **plugins\Log4jLoggingPlugin_2.0.0\Log4jLoggingPluginNodeUDN.class**
It seems that this Java class uses value **false** for **configurable** parameters when constructing the **NodeProperty** objects and would need to use value **true** to make the properties configurable.
 
Please don't hesitate to let me know if any further information is required.

Thank you so much in advance!

**Comment by nieder-github** [comment]
Thanks for reporting this. But I'm not sure I understand this completely.
You're right in what you are describing. Nevertheless I'm able to change the node for log4j properties in the bar file when deploying in my environment ( v12.0.3.0 ).
What are you intending to do ? Which version are you using ?

**Comment by lyes-aknouche** [comment]
Thank you for your quick reply and apologies for the confusion. Please see below further information.

I'm referring here to the "configurable properties" which can be overridden in a BAR file when using the `mqsiapplybaroverride` command.
In version 2.0.0 / 2.0.1, there are no "configurable properties" available to override for the log4j node used in the BAR file.
However, in version 1.1 / 1.2, all 4 properties of a log4j node were "configurable properties" and could be overridden using the `mqsiapplybaroverride` command.

A scenario where "configurable properties" is useful for the log4j node is when each ACE environment has a different path to the log4j config file and require to use the `mqsiapplybaroverride` before each deployment to define the correct path for the log4j config file.

Please see below how to reproduce the issue using release 2.0.1 and what is expected when using release 1.1.

I create a github repositories with the below files: https://github.com/lyes-aknouche/ace-log4j-demo-issue/tree/main/workspace/demoapp

### using release 2.0.1
**Environment**
- IBM App Connect Enterprise 12.0.7.0 on Windows
- Plugin `Log4jLoggingNode_v2.0.1.par` installed in ACE runtime D:\IBM\ACE\12.0.7.0\server\jplugin
- Plugin `Log4jLoggingPluginFeature_v2.0.0.zip` installed in ACE toolkit (please note that the PDF in release 2.0.1 indicates to use Log4jLoggingPluginFeature_v2.0.**1**.zip but this file is not available for download in github)

**Application (demoapp)**
- Simple message flow (demoflow.msgflow) with 1 MQ Input node receiving message, 1 log4j node writing some log and 1 MQ output node
- log4j node has the following properties defined in the toolkit:
```
   - Log4jConfigFile = brokerlog.xml
   - LogLevel = INFO
   - LogText = My Message is $ESQL:CAST(InputRoot.BLOB.BLOB AS CHARACTER CCSID 1252)$
   - LoggerName = FILE
```

**BAR file**
- The BAR file created is `demoapp-logging-plugin20.bar` which contains the demoapp application
- When using the mqsireadbar command to list all configurable properties, no entries are listed for the log4j node:
```
mqsireadbar -b D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin20.bar -r
BIP1052I: Reading Bar file using runtime mqsireadbar...
D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin20.bar:
  demoapp.appzip (6/03/23 10:39 AM):
    application.descriptor (6/03/23 10:39 AM):
    brokerlog-log4j.xml (6/03/23 10:39 AM):
    brokerlog-log4j2.xml (6/03/23 10:39 AM):
    demoflow.cmf (6/03/23 10:39 AM):
    Deployment descriptor:
      startMode
      demoflow#additionalInstances
      demoflow#commitCount
      demoflow#commitInterval
      demoflow#coordinatedTransaction
      demoflow#consumerPolicySet
      demoflow#providerPolicySet
      demoflow#consumerPolicySetBindings
      demoflow#providerPolicySetBindings
      demoflow#securityProfileName
      demoflow#monitoringProfile
      demoflow#startMode
      demoflow#startInstancesWhenFlowStarts
      demoflow#maximumRateMsgsPerSec
      demoflow#wlmPolicy
      demoflow#notificationThresholdMsgsPerSec
      demoflow#processingTimeoutSec
      demoflow#processingTimeoutAction
      demoflow#MQ Input.SSLCertificateLabel
      demoflow#MQ Input.SSLCipherSpec
      demoflow#MQ Input.SSLPeerName
      demoflow#MQ Input.additionalInstances
      demoflow#MQ Input.channelName
      demoflow#MQ Input.componentLevel
      demoflow#MQ Input.connection
      demoflow#MQ Input.destinationQueueManagerName
      demoflow#MQ Input.listenerPortNumber
      demoflow#MQ Input.policyUrl
      demoflow#MQ Input.queueManagerHostname
      demoflow#MQ Input.queueName = Q1
      demoflow#MQ Input.resetBrowseTimeout
      demoflow#MQ Input.securityIdentity
      demoflow#MQ Input.securityProfileName
      demoflow#MQ Input.serializationToken
      demoflow#MQ Input.topicProperty
      demoflow#MQ Input.useSSL
      demoflow#MQ Input.validateMaster
      demoflow#MQ Output.AddRequestToGroup
      demoflow#MQ Output.GroupRequestFolderName
      demoflow#MQ Output.GroupRequestTimeout
      demoflow#MQ Output.SSLCertificateLabel
      demoflow#MQ Output.SSLCipherSpec
      demoflow#MQ Output.SSLPeerName
      demoflow#MQ Output.channelName
      demoflow#MQ Output.connection
      demoflow#MQ Output.destinationQueueManagerName
      demoflow#MQ Output.listenerPortNumber
      demoflow#MQ Output.policyUrl
      demoflow#MQ Output.queueManagerHostname
      demoflow#MQ Output.queueManagerName
      demoflow#MQ Output.queueName = Q2
      demoflow#MQ Output.replyToQ
      demoflow#MQ Output.replyToQMgr
      demoflow#MQ Output.securityIdentity
      demoflow#MQ Output.securityProfileName
      demoflow#MQ Output.useSSL
      demoflow#MQ Output.validateMaster

BIP8071I: Successful command completion.
```
- The properties file used to override the properties is `demoapp-logging-plugin20.properties` which defines the full path where to find the log4j2 config file:
`demoflow#Log Message.log4jConfigFile = D:\git\github\ace-log4j-demo-issue\workspace\demoapp\brokerlog-log4j2.xml`
- The properties are overridden using the `mqsiapplybaroverride` command:
`mqsiapplybaroverride -b D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin20.bar -p D:\git\github\ace-log4j-demo-issue\workspace\demoapp\demoapp-logging-plugin20.properties -o D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin20-output.bar -r`

**Deployment**
- The updated bar file `demoapp-logging-plugin20-output.bar` is deployed using the below command:
`mqsideploy ACENODE -e acedemois -a D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin20-output.bar -m`

**Testing**
- when putting message in the input queue, the message is put in the output queue but no messages are logged in the log file as defined in the log4j2 config file.
- when starting the trace, it shows that the message flows tried to find log4j2 config file `brokerlog.xml` in its classpath which does not exist

### using plugin 1.2.4

**Environment**
- IBM App Connect Enterprise 12.0.7.0 on Windows
- Plugin file `Log4jLoggingNode_v1.2.4.jar` installed in ACE runtime D:\IBM\ACE\12.0.7.0\server\jplugin
- Plugin files `Log4jLoggingNode_v1.2.4.jar`, `jakarta-oro-2.0.4.jar` and `log4j-1.2.8.jar` installed in ACE runtime C:\ProgramData\IBM\MQSI\shared-classes
- Plugin `Log4jLoggingPluginFeature_v1.1.zip` installed in ACE toolkit

**Application (demoapp) - same as above**
- Simple message flow (demoflow.msgflow) with 1 MQ Input node receiving message, 1 log4j node writing some log and 1 MQ output node
- log4j node has the following properties defined in the toolkit:
```
   - Log4jConfigFile = brokerlog.xml
   - LogLevel = INFO
   - LogText = My Message is $ESQL:CAST(InputRoot.BLOB.BLOB AS CHARACTER CCSID 1252)$
   - LoggerName = FILE
```

**BAR file**
- BAR file created is `demoapp-logging-plugin11.bar` which contains the demoapp application
- When using the mqsireadbar command to list all configurable properties, all 4 configurable properties for the log4j node are listed:
```
D:\git\github\ace-log4j-demo-issue>mqsireadbar -b D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin11.bar -r
BIP1052I: Reading Bar file using runtime mqsireadbar...
D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin11.bar:
  demoapp.appzip (6/03/23 10:12 AM):
    application.descriptor (6/03/23 10:12 AM):
    brokerlog-log4j.xml (6/03/23 10:12 AM):
    brokerlog-log4j2.xml (6/03/23 10:12 AM):
    demoflow.cmf (6/03/23 10:12 AM):
    Deployment descriptor:
      startMode
      demoflow#additionalInstances
      demoflow#commitCount
      demoflow#commitInterval
      demoflow#coordinatedTransaction
      demoflow#consumerPolicySet
      demoflow#providerPolicySet
      demoflow#consumerPolicySetBindings
      demoflow#providerPolicySetBindings
      demoflow#securityProfileName
      demoflow#monitoringProfile
      demoflow#startMode
      demoflow#startInstancesWhenFlowStarts
      demoflow#maximumRateMsgsPerSec
      demoflow#wlmPolicy
      demoflow#notificationThresholdMsgsPerSec
      demoflow#processingTimeoutSec
      demoflow#processingTimeoutAction
      demoflow#Log Message.log4jConfigFile = brokerlog.xml
      demoflow#Log Message.logLevel = INFO
      demoflow#Log Message.logText = My Message is $ESQL:CAST(InputRoot.BLOB.BLOB AS CHARACTER CCSID 1252)$
      demoflow#Log Message.loggerName = FILE
      demoflow#MQ Input.SSLCertificateLabel
      demoflow#MQ Input.SSLCipherSpec
      demoflow#MQ Input.SSLPeerName
      demoflow#MQ Input.additionalInstances
      demoflow#MQ Input.channelName
      demoflow#MQ Input.componentLevel
      demoflow#MQ Input.connection
      demoflow#MQ Input.destinationQueueManagerName
      demoflow#MQ Input.listenerPortNumber
      demoflow#MQ Input.policyUrl
      demoflow#MQ Input.queueManagerHostname
      demoflow#MQ Input.queueName = Q1
      demoflow#MQ Input.resetBrowseTimeout
      demoflow#MQ Input.securityIdentity
      demoflow#MQ Input.securityProfileName
      demoflow#MQ Input.serializationToken
      demoflow#MQ Input.topicProperty
      demoflow#MQ Input.useSSL
      demoflow#MQ Input.validateMaster
      demoflow#MQ Output.AddRequestToGroup
      demoflow#MQ Output.GroupRequestFolderName
      demoflow#MQ Output.GroupRequestTimeout
      demoflow#MQ Output.SSLCertificateLabel
      demoflow#MQ Output.SSLCipherSpec
      demoflow#MQ Output.SSLPeerName
      demoflow#MQ Output.channelName
      demoflow#MQ Output.connection
      demoflow#MQ Output.destinationQueueManagerName
      demoflow#MQ Output.listenerPortNumber
      demoflow#MQ Output.policyUrl
      demoflow#MQ Output.queueManagerHostname
      demoflow#MQ Output.queueManagerName
      demoflow#MQ Output.queueName = Q2
      demoflow#MQ Output.replyToQ
      demoflow#MQ Output.replyToQMgr
      demoflow#MQ Output.securityIdentity
      demoflow#MQ Output.securityProfileName
      demoflow#MQ Output.useSSL
      demoflow#MQ Output.validateMaster

BIP8071I: Successful command completion.
```
- The properties file used to override the properties is `demoapp-logging-plugin11.properties` which defines the full path where to find the log4j1 config file:
`demoflow#Log Message.log4jConfigFile = D:\git\github\ace-log4j-demo-issue\workspace\demoapp\brokerlog-log4j.xml`
- The properties are overridden using the `mqsiapplybaroverride` command:
`mqsiapplybaroverride -b D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin11.bar -p D:\git\github\ace-log4j-demo-issue\workspace\demoapp\demoapp.properties -o D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin11-output.bar -r
`

**Deployment**
- The updated bar file `demoapp-logging-plugin11-output.bar` is deployed using the below command:
`mqsideploy ACENODE -e acedemois -a D:\git\github\ace-log4j-demo-issue\workspace\BARfiles\demoapp-logging-plugin11-output.bar -m`

**Testing**
- when putting message in the input queue, the message logged in the expected log file defined in the log4j config file C:/temp/Log4jTest-log4j.log:
```
10:13:16,698 INFO  [acedemois.demoflow] My Message is <test>abc</test>
10:13:57,471 INFO  [acedemois.demoflow] My Message is test1
```

Please note that the above tests were also done in the Linux environment with the same results.

### Expectation
- the expectation is when using release 2.0.0 / 2.0.1, "configurable properties" for any log4j node should be listed as it was the case with release 1.1 / 1.2

### Resolution
it seems that there are 2 changes required in the Log4jLoggingPluginFeature_v2.0.0.zip

1. `plugins\Log4jLoggingPlugin_2.0.0\Log4jLoggingPlugin.msgnode`
    This file used to have (in versions 1.1 / 1.2) the attributes `configurable="true"` for each XML element `propertyDescriptor`
2. `plugins\Log4jLoggingPlugin_2.0.0\Log4jLoggingPluginNodeUDN.class`
    It seems that this Java implementation class uses value `false` for configurable parameters when constructing the `NodeProperty `objects and would need to use value true to make the properties configurable.

Apologies for the long post.

Please don't hesitate to let me know if any further information is required.

**Comment by nieder-github** [comment]
I just released v2.0.2 to fix this issue.
